### PR TITLE
DMAS Phase 1 Work

### PR DESCRIPTION
**Pull request by BabarKhanDev**
Human Detection Model
Plant Identification
PostgreSQL Code
/upload_image 
Fake Data Generation

### CODE REVIEW

File: Models/HumanDetection/HumanDetector.py [hunk 1]
```diff
@@ -205,7 +205,7 @@ def display_training_graphs(src_location: str):
 
 if __name__ == "__main__":
```
Comment (cjd552) [inline Models/HumanDetection/HumanDetector.py:206]: The training looks pretty reasonable; for the demo we would probably have it pretrained and need a test/inference method to be called, right? Is that stored somewhere else or not written yet?

For the accuracy we can always just find some images that it outputs the correct result on, and make sure to use those in the demo to ensure that it exhibits the correct behaviour.

Comment (BabarKhanDev) [reply Models/HumanDetection/HumanDetector.py:206]: I have a trained model stored on my laptop, I can just give the weights to whoever is running the demo on a usb stick. 

testing/inference is all stored here, you can see how it's done in app/main.py

That's a good idea, I just have human detection disabled by default in main.py but if we want to show it then yeah lets do that :)

File: Util/Data.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    date: datetime
+
+
+class ProcessedEntry(BaseModel):
+    image_uri: str
+    plant_id: str
+# Database entries end
+
+
+def load_config(section='postgresql', filename='database.ini'):
```
Comment (Laura7089) [inline Util/Data.py:22]: Is this the only way to configure the database connection?

File: Util/FileHandler.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+from PIL.ExifTags import GPSTAGS, TAGS
+
+
+def save(image: Image, filename: str) -> str:
+    # TODO replace with S3
+    if not os.path.exists('Data/'):
+        os.makedirs('Data')
+
+    _, ext = filename.split(".")
+    filename = uuid.uuid4().hex
```
Comment (Laura7089) [inline Util/FileHandler.py:14]: I see the utility in using a UUID, but we could avoid storing identical images by using a hash. What's the number of images we're typically ingesting here?

File: Util/FileHandler.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+
+def gps_details(image: Image) -> {}:
+    exif_table = {}
+    info = image.getexif()
+    for tag, value in info.items():
+        decoded = TAGS.get(tag, tag)
+        exif_table[decoded] = value
+
+    if 'GPSInfo' not in exif_table.keys():
```
Comment (Laura7089) [inline Util/FileHandler.py:26]: Anonymous plants get sent to [Null Island](https://en.wikipedia.org/wiki/Null_Island), nice.

Comment (Laura7089) [reply Util/FileHandler.py:26]: Perhaps it should fail if there's no attached coordinates, since in that case the image isn't useful to us?

File: Util/PlantDetector.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+
+def detect(image_path: str):
+    with open("plantnet_API_key", "r") as file:
+        API_KEY = file.readline()
+
+    PROJECT = "all"  # try "weurope" or "canada"
+    api_endpoint = f"https://my-api.plantnet.org/v2/identify/{PROJECT}?api-key={API_KEY}"
+
+    data = {'organs': ['flower']}
```
Comment (Laura7089) [inline Util/PlantDetector.py:15]: > Organs associated to images (must contains one of: leaf, flower, fruit, bark, auto. Could contains: habit, other)

I'm not totally sure what this does.
Does this mean we only ID the first plant with a flower?

File: Util/PlantDetector.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    ]
+
+    req = requests.Request('POST', url=api_endpoint, files=files, data=data)
+    prepared = req.prepare()
+
+    s = requests.Session()
+    response = s.send(prepared)
+    json_result = json.loads(response.text)
+    if 'statusCode' in json_result.keys() and json_result['statusCode'] == 404:
+        raise PlantsUndetectedError("Plant not recognised")
```
Comment (Laura7089) [inline Util/PlantDetector.py:28]: Side note: this is a terrible way to use HTTP status codes, imo - shame on you PlantNet

File: app/main.py [hunk 7]
```diff
@@ -1,24 +1,112 @@
...
+
+from Exceptions.PlantsUndetectedError import PlantsUndetectedError
+from Util import FileHandler
+from Util.Data import RawEntry, ProcessedEntry, load_config, connect, insert_raw_entry, insert_processed_entry
+from Models.HumanDetection.HumanDetector import Classifier as HumanDetector
+from Util.PlantDetector import detect
+from app.Messages import StatusEnum, MessageResponse, PlantGrowthDataResponse
+
+# START OPTIONS
+CHECK_HUMANS = False  # Verify if humans are in the image
```
Comment (Laura7089) [inline app/main.py:17]: Should we leave a TODO here?

File: app/main.py [hunk 8]
```diff
@@ -1,24 +1,112 @@
...
+        file.file.close()
+        buf.close()
+
+    # Check For Humans
+    img_normalised = human_detector.tensor_transform(im)
+    img_normalised = img_normalised.unsqueeze_(0)
+    img_normalised = img_normalised.to(device)
+    output = human_detector(img_normalised)
+    output = torch.argmax(output.detach().cpu())
+    if output == 1 and CHECK_HUMANS:
```
Comment (Laura7089) [inline app/main.py:71]: Can save time and compute power by reading `CHECK_HUMANS` before we do any of this

**Review by Laura7089 — CHANGES_REQUESTED** [review]
I don't *really* follow all the machine learning stuff you're doing here; it's not my wheelhouse.

File: Models/HumanDetection/HumanDetector.py [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+        self.classes = ["No People", "Yes People"]
+        self.tensor_transform = transforms.Compose([
+            transforms.PILToTensor(),
+            transforms.Resize((128, 128), antialias=True),
+            transforms.Lambda(lambda y: y.float() / 255),
+            transforms.Normalize(0.5, 0.5),
+        ])
+
+        # Define Model
+        self.conv_layers = nn.Sequential(
```
[the remark this answers is not in the document]
Comment (BabarKhanDev) [reply Models/HumanDetection/HumanDetector.py:34]: I made it quite quickly as something that worked as an MVP for the demo but it's definitely not the best, probably not even close haha 😅 

If we have time closer to the demo date I'll spend some time improving it

File: .gitignore [hunk 10]
```diff
@@ -0,0 +1,5 @@
+/Models/HumanDetection/weights/
+/Models/HumanDetection/Datasets/
+/Data/
+plantnet_API_key
+database.ini
```
[the remark this answers is not in the document]
Comment (BabarKhanDev) [reply .gitignore:5]: Ah I didn't know that ty, seen as we only run the database locally I'll just remove this from the gitignore

File: Models/HumanDetection/HumanDetector.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Laura7089) [reply Models/HumanDetection/HumanDetector.py:34]: Fair enough, as long as it works well enough :heart:

File: SQL/init.sql [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    image_uri varchar COLLATE pg_catalog."default" NOT NULL,
+    plant_id varchar COLLATE pg_catalog."default",
+    CONSTRAINT processed_entry_pkey PRIMARY KEY (id),
+    CONSTRAINT processed_entry_raw_entry_image_uri_fk FOREIGN KEY (image_uri)
+        REFERENCES image_processing.raw_entry (image_uri) MATCH SIMPLE
+        ON UPDATE NO ACTION
+        ON DELETE NO ACTION
+)
+
+TABLESPACE pg_default;
```
[the remark this answers is not in the document]
Comment (BabarKhanDev) [reply SQL/init.sql:21]: I just copied the commands that pgAdmin generated, I've just looked it up and it should use pg_default by default, will remove :)

File: SQL/init.sql [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    image_uri varchar COLLATE pg_catalog."default" NOT NULL,
+    latitude double precision[] NOT NULL,
+    longitude double precision[] NOT NULL,
+    date timestamp NOT NULL,
+    CONSTRAINT raw_entry_pkey PRIMARY KEY (image_uri)
+)
+
+TABLESPACE pg_default;
+
+ALTER TABLE IF EXISTS image_processing.raw_entry
```
[the remark this answers is not in the document]
Comment (BabarKhanDev) [reply SQL/init.sql:40]: Sounds good w me

File: Util/Data.py (hunk 2 above)
Comment (BabarKhanDev) [reply Util/Data.py:22]: We can just run conn = psycopg2.connect(username=, password=, host=, port=), that's all this function does

File: Util/FileHandler.py (hunk 3 above)
Comment (BabarKhanDev) [reply Util/FileHandler.py:14]: That's a good idea, I was mostly concerned with name collisions rather than identical content. I'm not sure how many images we expect to ingest, I don't think we've considered that yet.

File: Util/FileHandler.py (hunk 4 above)
Comment (BabarKhanDev) [reply Util/FileHandler.py:26]: That's a shout, will update

File: Util/PlantDetector.py (hunk 5 above)
Comment (BabarKhanDev) [reply Util/PlantDetector.py:15]: I think it means that the API is looking for flowers in the image. 

When I update the processing to store results that pass a threshold I can also update this to also check for leafs, fruit, and bark

The API sends back a list of results saying what it thinks the image contains, I think it's designed for 1 plant but I can't see why it doesn't work with multiple, I'll also do some testing for that

File: Util/PlantDetector.py (hunk 6 above)
Comment (BabarKhanDev) [reply Util/PlantDetector.py:28]: haha yeah it doesn't send back any status code if it's successful 🤦🏼

File: app/main.py (hunk 8 above)
Comment (BabarKhanDev) [reply app/main.py:71]: That's a great idea, updated

File: .gitignore (hunk 10 above)
[the remark this answers is not in the document]
Comment (BabarKhanDev) [reply .gitignore]: ooh thank you, I didn't know that. Added :)

File: Util/PlantDetector.py (hunk 5 above)
Comment (Laura7089) [reply Util/PlantDetector.py:16]: Worth a TODO note then?

File: .gitignore (hunk 10 above)
[the remark this answers is not in the document]
Comment (Laura7089) [reply .gitignore]: Sorry I didn't say you then keep the ignore line for the "real" file in the `.gitignore`.

[the remark this answers is not in the document]
Comment (Laura7089) [reply .gitignore]: Bump on the above

File: .gitignore
Comment (Laura7089) [inline .gitignore:1]: You seem to be missing `psycopg2` from the requirements.

**Review by Laura7089 — COMMENTED** [review]
Type hinting is cool, but currently `mypy` fails to resolve anything because we're pulling in so many libraries which aren't typed, and so unfortunately the hinting is unverifiable.

File: .gitignore
Comment (BabarKhanDev) [reply .gitignore:1]: Ty :)

File: Util/PlantDetector.py (hunk 5 above)
Comment (BabarKhanDev) [reply Util/PlantDetector.py:16]: Added

**Review by Laura7089 — APPROVED** [review]
